NETWORK_PROTOCOL.md

```diff
@@ -1 +1,102 @@
-Placeholder for network protocol.
+Status: this is a rough draft, intended to get some ideas out there. 
+
+Missing: how to run on top of HTTP/2. Possibly the structure or semantics of publisher name strings.
+Might be removed: extension support.
+Need to choose serialization method (protobuf, msgpack, other?)
+
+## Transport assumptions
+
+A supporting transport must be similar to a streaming socket:
+
+1. Bidirectional and full-duplex
+2. An octet stream, i.e. all octet values may be sent unencoded
+3. Ordered delivery: an implementation may map protocol messages to some features of the underlying transport (e.g. [Ømq](http://zeromq.org/) messages), but the messages must arrive in the same order as they were sent
+
+Definitely supported transports include TCP, TLS (over TCP), WebSockets, and most socket-like objects (e.g. pipes). HTTP/2 will be supported, but may require a dedicated specification for implementing this protocol.
+
+## Message framing
+
+An existing serialization format should be used. Current candidates are [Protocol Buffers](https://github.com/google/protobuf/) (which is slightly less space efficient), [MessagePack](http://msgpack.org/) (whose Scala implementation may not be as good as the others), and possibly [Thrift](https://thrift.apache.org/) (with which I'm less familiar). 
```

Comment:
@danarmak SBE will be adding a JavaScript codec in the future. Or so a little bird tells me...

The needs of efficient (in length, processing, etc.) data layout/serialization are complex tradeoffs. And by and large, I think they are orthogonal. However, I think we have to at least have a place to start and a default fallback.

GPB (latest version from gRPC) is OK for most uses and a sensible fallback.

However, I would add the ability in the setup of the session to specify the serialization mechanic and thus allow users to pick what they want. It isn't one size fits all use cases. So, it needs to be possible to slide in your own somehow.